docs/kb/edk2-capsule-updates.md

```diff
@@ -325,6 +328,326 @@ interferes).
 
 [wiki-pkc]: https://en.wikipedia.org/wiki/Public-key_cryptography
 
+## Generating signing keys with OpenSSL
+
+The process involves creation of 3 certificates, a local certificate
+authority (CA) in a directory and 14 files.  This warrants some overview of
+what's going to be done.
+
+Keys making up the chain will assume the following roles:
+
+- _root_ — self-signed certificate acting as the basis of a CA
+- _sub_ — second-level CA signed by _root_ key
+- _sign_ — a signing key (not a CA) that is signed by _sub_ key
+
+A digital signature algorithm (DSA) consists of a private key (PK) algorithm and
+a digest algorithm.  Keys in a chain don't have to use the same DSA.  Also, most
+DSAs allow use of an arbitrary digest (not Ed25519 or Ed448 which prescribe the
+use of SHA-512 and SHAKE-256 respectively).
+
+Keys exist independently of the chains in which they appear.  When stored,
+private keys are often encrypted.
+
+### Key generation by example
+
+It will be easier to go into finer details while looking at the actual commands.
+The entire process can be broken into several stages:
+
+1. Creation of 3 certificates
+2. Construction of a CA
+3. Preparation of _sign_ certificate for signing
+4. Preparation of _root_ certificate for inclusion into EDK
+
+#### Make certificates
+
+Using 4096-bit RSA keys as an example:
+
+```bash
+openssl genrsa -aes256 -out root.p8e 4096
+openssl genrsa -aes256 -out sub.p8e 4096
+openssl genrsa -aes256 -out sign.p8e 4096
+```
+
+Each of them is encrypted with AES-256.  The password is queried interactively
+for creation and each time the certificate is used to sign something.  Drop
+`-aes256` to not use encryption (for a test or if you consider access to the
+files a complete compromise of the security).
+
+`.p8e` extension is for PKCS #8 format carrying an encrypted private key.
+
+#### Make a CA
+
+By default, directory for a CA is called `demoCA` (can be changed in
+`/etc/ssl/openssl.cnf`).  The directory needs to be set up prior to using
+`openssl ca`:
+
+```bash
+mkdir -p demoCA/newcerts
+touch demoCA/index.txt
+echo 01 > demoCA/serial
+```
+
+Initialize it with self-signed root certificate (will ask for a password
+and certificate fields; country, state and organization fields must match _root_
+certificate, common name must be a unique value):
+
+```bash
+openssl req -new -x509 -days 3650 -key root.p8e -out root.pub.pem
+```
+
+Create certificate signing requests (CSRs) (don't bother with
+entering a challenge for CSRs, you won't be asked for it):
+
+```bash
+openssl req -new -key sub.p8e -out sub.csr
+openssl req -new -key sign.p8e -out sign.csr
+```
+
+Perform the signing (there will be password and confirmation prompts):
+
+```bash
+openssl ca -extensions v3_ca \
+           -in sub.csr \
+           -days 3650 \
+           -cert root.pub.pem \
+           -keyfile root.p8e \
+           -notext \
```

Comment:
Is there a reason why `-notext` is here but not for signing certificate? If not, it would be better to be consistent and either use it in both cases or neither of them.

Reply:
https://github.com/Dasharo/docs/compare/c791ff82210eed388fe9a65c78a80490c8d13b24..fcda915b6e88ce72089bfc5a43900ac61336c7ba#diff-b0d92435e1999e660dd9b88bd5bd5dd14331f32b3abce08c0b9a5895537af576R420-R431